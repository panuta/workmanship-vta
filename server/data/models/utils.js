
// TODO : bulkUpdate base on some key

export const parseDuration = (text) => {
  const [hoursText, minutesText] = text.split('.')
  try {
    return parseInt(hoursText, 10) * 60 + parseInt(minutesText, 10)
  } catch (error) {
    return null
  }
}
