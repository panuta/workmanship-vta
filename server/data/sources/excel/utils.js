import _ from 'lodash'
import XLSX from 'xlsx'

export const readColumnData = (worksheet, range, parser) => {
  const { s: { c: startCol, r: startRow }, e: { r: endRow } } = XLSX.utils.decode_range(range)
  return _.range(startRow, endRow + 1)
    .map(row => {
      const cell = worksheet[XLSX.utils.encode_cell({ c: startCol, r: row })]
      if(cell === undefined || cell.v === undefined) return null

      if(_.isFunction(parser)) {
        return parser(cell)
      }
      return cell.v
    })
}

export const readColumnsData = (worksheet, columns) => {
  const data = columns.map(column => readColumnData(worksheet, column.range, column.parser))
  return _.zip(...data).map(datum => _.zipObject(columns.map(column => column.key), datum))
}

export const columnRange = (start, end) => {
  return _.range(XLSX.utils.decode_col(start), XLSX.utils.decode_col(end) + 1)
    .map(colIndex => XLSX.utils.encode_col(colIndex))
}
