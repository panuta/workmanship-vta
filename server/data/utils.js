import dayjs from 'dayjs'

export const parseDuration = (text) => {
  const [hoursText, minutesText] = text.split('.')
  try {
    return parseInt(hoursText, 10) * 60 + parseInt(minutesText, 10)
  } catch (error) {
    return null
  }
}

export const dateToString = date => {
  return dayjs(date).format('YYYY-MM-DD')
}
