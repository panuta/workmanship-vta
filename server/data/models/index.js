import _ from 'lodash'
import dayjs from 'dayjs'
import { Op } from 'sequelize'
import { Employee, EmployeeAttendance, Shift } from './definitions'
import { dateToString } from '../utils'

export const initDatabase = async () => {
  await Shift.sync({ alter: false })
  await Employee.sync({ alter: false })
  await EmployeeAttendance.sync({ alter: false })
}

export const findEmployees = async (activeInMonth) => {
  if(_.isDate(activeInMonth)) {
    return Employee.findAll({
      where: {
        terminationDate: {
          [Op.or]: {
            [Op.is]: null,
            [Op.gte]: dateToString(new Date(activeInMonth.getFullYear(), activeInMonth.getMonth(), 1)),
          }
        }
      }
    })
  }
  return Employee.findAll()
}

export const findEmployeeAttendances = async (attendanceMonth) => {
  const date = dayjs(attendanceMonth)
  const attendanceStart = date.startOf('month').toDate()
  const attendanceEnd = date.endOf('month').toDate()

  return EmployeeAttendance.findAll({
    where: {
      attendanceDate: {
        [Op.between]: [attendanceStart, attendanceEnd]
      }
    }
  })
}
