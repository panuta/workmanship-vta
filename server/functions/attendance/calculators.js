import _ from 'lodash'
import { isUntilAttendanceMonth, sameAttendanceMonth } from '../../utils/attendanceMonth'

export const increaseByValueUntilAttendanceMonth = (attendance, currentValue, attendanceMonth, valueKey) => {
  // eslint-disable-next-line no-param-reassign
  if(_.isNil(currentValue)) currentValue = 0

  const value = attendance[valueKey]
  if(_.isNumber(value) && isUntilAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + value
  }
  return currentValue
}

export const increaseByValueOnAttendanceMonth = (attendance, currentValue, attendanceMonth, valueKey) => {
  // eslint-disable-next-line no-param-reassign
  if(_.isNil(currentValue)) currentValue = 0

  const value = attendance[valueKey]
  if(_.isNumber(value) && sameAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + value
  }
  return currentValue
}

export const increaseWhenShiftMatchedUntilAttendanceMonth = (attendance, currentValue, attendanceMonth, shiftNames) => {
  // eslint-disable-next-line no-param-reassign
  if(_.isNil(currentValue)) currentValue = 0

  if(shiftNames.includes(attendance.shift) && isUntilAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + 1
  }
  return currentValue
}

/**
 * Increase value by 1 when attendance shift is one of shiftNames
 *
 * @param attendance => Employee attendance data
 * @param shiftNames: Array<String>
 * @param currentValue: Integer
 * @param attendanceMonth
 * @return {*}
 */
export const increaseWhenShiftMatchedOnAttendanceMonth = (attendance, currentValue, attendanceMonth, shiftNames) => {
  // eslint-disable-next-line no-param-reassign
  if(_.isNil(currentValue)) currentValue = 0

  if(shiftNames.includes(attendance.shift) && sameAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + 1
  }
  return currentValue
}

/**
 * Increase value by 1 when attendance shift is NOT one of shiftNames
 *
 * @param attendance => Employee attendance data
 * @param shiftNames: Array<String>
 * @param currentValue: Integer
 * @param attendanceMonth
 * @return {*}
 */
export const increaseWhenShiftNotMatchedOnAttendanceMonth = (attendance, currentValue, attendanceMonth, shiftNames) => {
  // eslint-disable-next-line no-param-reassign
  if(_.isNil(currentValue)) currentValue = 0

  if(!shiftNames.includes(attendance.shift) && sameAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + 1
  }
  return currentValue
}
