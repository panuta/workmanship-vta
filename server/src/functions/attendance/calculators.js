import { isNil } from 'lodash'
import { sameAttendanceMonth } from '../../utils/attendanceMonth'
import { parseInteger } from '../../utils/dataType'

export const increaseByValue = (attendance, currentValue, attendanceMonth, valueKey) => {
  const value = parseInteger(attendance[valueKey], 0)
  return !isNil(currentValue) ? currentValue + value : value
}

export const increaseByValueOnAttendanceMonth = (attendance, currentValue, attendanceMonth, valueKey) => {
  if(sameAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    const value = parseInteger(attendance[valueKey], 0)
    return !isNil(currentValue) ? currentValue + value : value
  }
  return !isNil(currentValue) ? currentValue : 0
}

export const increaseWhenShiftMatched = (attendance, currentValue, attendanceMonth, shiftNames) => {
  // eslint-disable-next-line no-param-reassign
  if(isNil(currentValue)) currentValue = 0

  if(shiftNames.includes(attendance.shift)) {
    return currentValue + 1
  }
  return currentValue
}

/**
 * Increase value by 1 when attendance shift is one of shiftNames
 *
 * @param attendance => Employee attendance data
 * @param shiftNames: Array<String>
 * @param currentValue: Integer
 * @param attendanceMonth
 * @return {*}
 */
export const increaseWhenShiftMatchedOnAttendanceMonth = (attendance, currentValue, attendanceMonth, shiftNames) => {
  // eslint-disable-next-line no-param-reassign
  if(isNil(currentValue)) currentValue = 0

  if(shiftNames.includes(attendance.shift) && sameAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + 1
  }
  return currentValue
}

/**
 * Increase value by 1 when attendance shift is NOT one of shiftNames
 *
 * @param attendance => Employee attendance data
 * @param shiftNames: Array<String>
 * @param currentValue: Integer
 * @param attendanceMonth
 * @return {*}
 */
export const increaseWhenShiftNotMatchedOnAttendanceMonth = (attendance, currentValue, attendanceMonth, shiftNames) => {
  // eslint-disable-next-line no-param-reassign
  if(isNil(currentValue)) currentValue = 0

  if(!shiftNames.includes(attendance.shift) && sameAttendanceMonth(attendance.attendanceDate, attendanceMonth)) {
    return currentValue + 1
  }
  return currentValue
}
