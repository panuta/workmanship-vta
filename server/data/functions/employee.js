import { Op } from 'sequelize'
import { Employee } from '../models'
import { dateToString, monthPeriod } from '../utils'

export const findEmployees = async (attendanceMonth) => {
  if(attendanceMonth !== undefined) {
    /**
     * Active employee
     * ( startDate is before end of month period )
     * AND
     * ( ( terminationDate is null ) OR ( terminationDate is after start of period ) )
     */
    const [startPeriod, endPeriod] = monthPeriod(attendanceMonth.year(), attendanceMonth.month())
    const startPeriodString = dateToString(startPeriod)
    const endPeriodString = dateToString(endPeriod)
    return Employee.findAll({
      where: {
        [Op.and]: [
          { startDate: {
            [Op.lte]: endPeriodString } },
          { terminationDate: {
            [Op.or]: [
              { [Op.is]: null },
              { [Op.gte]: startPeriodString },
            ] } },
        ]
      }
    })
  }
  return Employee.findAll()
}
