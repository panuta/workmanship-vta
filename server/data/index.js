import { findEmployeeAttendances, findEmployees } from './models'

export const getMonthlyEmployeeAttendances = async (attendanceMonth) => {
  const employees = await findEmployees(attendanceMonth)
  const employeeAttendances = await findEmployeeAttendances(attendanceMonth)

  const employeeMapping = employees.reduce((accumulator, employee) => {
    accumulator[employee.code] = {
      employee,
      vacation: 0
    }
    return accumulator
  }, {})

  employeeAttendances.forEach(employeeAttendance => {
    // Vacation
    if(employeeAttendance.shift === 'พักร้อน') {
      employeeMapping[employeeAttendance.code].vacation += 1
    }


  })

  const result = Object.values(employeeMapping).map(employeeData => {
    const employeeDict = employeeData.employee.toJSON()
    // console.log(employeeDict)
    return Object.assign(employeeDict, {
      vacation: employeeData.vacation
    })
  })

  // console.log(result)
  return result
}
