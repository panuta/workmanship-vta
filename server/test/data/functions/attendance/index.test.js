import Config from '../../../../src/config'
import { initDatabase, EmployeeAttendance, closeDatabase, Employee } from '../../../../src/data/models'
import {
  calculateEmployeeAttendances,
  getEmployeesAttendancesUntilAttendanceMonth
} from '../../../../src/functions/attendance'
import { getEmployees } from '../../../../src/functions/employee'
import { toMomentObject, utcDate } from '../../../../src/utils/date'
import { increaseByValue } from '../../../../src/functions/attendance/calculators'

jest.mock('../../../../src/config')
jest.mock('../../../../src/functions/employee')

describe('Employee Attendance functions', () => {
  beforeAll(async () => {
    Config.__mockConfig({
      databaseStorage: ':memory:',
      databaseTimestamps: false
    })
    await initDatabase()
  })

  afterEach(async () => {
    jest.resetAllMocks()
    await EmployeeAttendance.destroy({ truncate: true, restartIdentity: true })
  })

  afterAll(async () => {
    await closeDatabase()
  })

  describe('getEmployeesAttendancesUntilAttendanceMonth function', () => {
    const setupEmployeesMockData = async () => {
      await EmployeeAttendance.bulkCreate([
        { code: '1001', attendanceDate: utcDate(2019, 11, 25), shift: 'A' },
        { code: '1001', attendanceDate: utcDate(2019, 11, 26), shift: 'A' },
        { code: '1001', attendanceDate: utcDate(2020, 8, 25), shift: 'A' },
        { code: '1001', attendanceDate: utcDate(2020, 8, 26), shift: 'A' },
        { code: '1001', attendanceDate: utcDate(2020, 11, 25), shift: 'A' },
        { code: '1001', attendanceDate: utcDate(2020, 11, 26), shift: 'A' },

        { code: '1002', attendanceDate: utcDate(2019, 11, 25), shift: 'A' },
        { code: '1002', attendanceDate: utcDate(2019, 11, 26), shift: 'A' },
        { code: '1002', attendanceDate: utcDate(2020, 8, 25), shift: 'A' },
        { code: '1002', attendanceDate: utcDate(2020, 8, 26), shift: 'A' },
        { code: '1002', attendanceDate: utcDate(2020, 11, 25), shift: 'A' },
        { code: '1002', attendanceDate: utcDate(2020, 11, 26), shift: 'A' },

        { code: '1003', attendanceDate: utcDate(2019, 11, 25), shift: 'A' },
        { code: '1003', attendanceDate: utcDate(2020, 11, 26), shift: 'A' },

        { code: '1004', attendanceDate: utcDate(2019, 11, 26), shift: 'A' },
        { code: '1004', attendanceDate: utcDate(2020, 11, 25), shift: 'A' },
      ])
    }

    it('should return from all employees', async () => {
      await setupEmployeesMockData()

      const attendanceMonth = toMomentObject(utcDate(2020, 8, 1))
      await expect(
        getEmployeesAttendancesUntilAttendanceMonth(attendanceMonth)
      ).resolves.toMatchSnapshot()
    })

    it('should return from selected employees', async () => {
      await setupEmployeesMockData()

      const attendanceMonth = toMomentObject(utcDate(2020, 8, 1))
      await expect(
        getEmployeesAttendancesUntilAttendanceMonth(attendanceMonth, ['1001', '1003'])
      ).resolves.toMatchSnapshot()
    })
  })


  describe('calculateEmployeeAttendances function', () => {
    it('should return attendances within a year for active employees', async () => {
      getEmployees.mockResolvedValue([
        Employee.build({ code: '1001' }),
        Employee.build({ code: '1002' }),
        Employee.build({ code: '1003' })
      ])

      await EmployeeAttendance.bulkCreate([
        { code: '1001', attendanceDate: utcDate(2019, 11, 25),
          shift: 'พักร้อน', notice: 1, compensation: 1, minutesLate: 10, minutesEarlyLeave: 20 },
        { code: '1001', attendanceDate: utcDate(2019, 11, 26),
          shift: 'A', notice: 1, compensation: 0, minutesLate: 5, minutesEarlyLeave: 3 },
        { code: '1001', attendanceDate: utcDate(2019, 11, 27),
          shift: 'ลากิจ', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1001', attendanceDate: utcDate(2019, 11, 28),
          shift: 'ลาป่วย', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1001', attendanceDate: utcDate(2019, 11, 29),
          shift: 'A', notice: 0, compensation: 1, minutesLate: 3, minutesEarlyLeave: 2 },
        { code: '1001', attendanceDate: utcDate(2020, 0, 1),
          shift: 'พักร้อน', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1001', attendanceDate: utcDate(2020, 0, 2),
          shift: 'ขาด', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 9 },
        { code: '1001', attendanceDate: utcDate(2020, 8, 3),
          shift: 'ลาป่วย', notice: 0, compensation: 1, minutesLate: 5, minutesEarlyLeave: 7 },
        { code: '1001', attendanceDate: utcDate(2020, 11, 25),
          shift: 'A', notice: 1, compensation: 0, minutesLate: 7, minutesEarlyLeave: 0 },
        { code: '1001', attendanceDate: utcDate(2020, 11, 26),
          shift: 'ลากิจ', notice: 1, compensation: 1, minutesLate: 10, minutesEarlyLeave: 5 },

        { code: '1002', attendanceDate: utcDate(2019, 11, 25),
          shift: 'พักร้อน', notice: 1, compensation: 1, minutesLate: 9, minutesEarlyLeave: 20 },
        { code: '1002', attendanceDate: utcDate(2019, 11, 26),
          shift: 'A', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2019, 11, 27),
          shift: 'ลากิจ', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2019, 11, 28),
          shift: 'ลาป่วย', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2019, 11, 29),
          shift: 'A', notice: 0, compensation: 1, minutesLate: 3, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2020, 0, 1),
          shift: 'A', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2020, 0, 2),
          shift: 'A', notice: 0, compensation: 0, minutesLate: 10, minutesEarlyLeave: 4 },
        { code: '1002', attendanceDate: utcDate(2020, 8, 3),
          shift: 'ลาป่วย', notice: 0, compensation: 1, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2020, 11, 25),
          shift: 'A', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
        { code: '1002', attendanceDate: utcDate(2020, 11, 26),
          shift: 'ลากิจ', notice: 1, compensation: 1, minutesLate: 10, minutesEarlyLeave: 5 },
      ])

      const attendanceMonth = toMomentObject(utcDate(2020, 8, 1))
      const values = await calculateEmployeeAttendances(attendanceMonth, [
        { fn: increaseByValue, resultKey: 'notice', args: ['notice'] }
      ])
      console.log(values)
    })
  })

  // describe('getMonthlyEmployeesAttendances function',  () => {
  //   beforeEach(() => {
  //     getEmployees.mockResolvedValue([
  //       Employee.build({ code: '1001' }),
  //       Employee.build({ code: '1002' }),
  //       Employee.build({ code: '1003' })
  //     ])
  //   })
  //
  //   it('should return correct values', async () => {
  //     await EmployeeAttendance.bulkCreate([
  //       { code: '1001', attendanceDate: new Date(2019, 11, 25),
  //         shift: 'พักร้อน', notice: 1, compensation: 1, minutesLate: 10, minutesEarlyLeave: 20 },
  //       { code: '1001', attendanceDate: new Date(2019, 11, 26),
  //         shift: 'A', notice: 1, compensation: 0, minutesLate: 5, minutesEarlyLeave: 3 },
  //       { code: '1001', attendanceDate: new Date(2019, 11, 27),
  //         shift: 'ลากิจ', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1001', attendanceDate: new Date(2019, 11, 28),
  //         shift: 'ลาป่วย', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1001', attendanceDate: new Date(2019, 11, 29),
  //         shift: 'A', notice: 0, compensation: 1, minutesLate: 3, minutesEarlyLeave: 2 },
  //       { code: '1001', attendanceDate: new Date(2020, 0, 1),
  //         shift: 'พักร้อน', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1001', attendanceDate: new Date(2020, 0, 2),
  //         shift: 'ขาด', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 9 },
  //       { code: '1001', attendanceDate: new Date(2020, 8, 3),
  //         shift: 'ลาป่วย', notice: 0, compensation: 1, minutesLate: 5, minutesEarlyLeave: 7 },
  //       { code: '1001', attendanceDate: new Date(2020, 11, 25),
  //         shift: 'A', notice: 1, compensation: 0, minutesLate: 7, minutesEarlyLeave: 0 },
  //       { code: '1001', attendanceDate: new Date(2020, 11, 26),
  //         shift: 'ลากิจ', notice: 1, compensation: 1, minutesLate: 10, minutesEarlyLeave: 5 },
  //
  //       { code: '1002', attendanceDate: new Date(2019, 11, 25),
  //         shift: 'พักร้อน', notice: 1, compensation: 1, minutesLate: 9, minutesEarlyLeave: 20 },
  //       { code: '1002', attendanceDate: new Date(2019, 11, 26),
  //         shift: 'A', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2019, 11, 27),
  //         shift: 'ลากิจ', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2019, 11, 28),
  //         shift: 'ลาป่วย', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2019, 11, 29),
  //         shift: 'A', notice: 0, compensation: 1, minutesLate: 3, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2020, 0, 1),
  //         shift: 'A', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2020, 0, 2),
  //         shift: 'A', notice: 0, compensation: 0, minutesLate: 10, minutesEarlyLeave: 4 },
  //       { code: '1002', attendanceDate: new Date(2020, 8, 3),
  //         shift: 'ลาป่วย', notice: 0, compensation: 1, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2020, 11, 25),
  //         shift: 'A', notice: 0, compensation: 0, minutesLate: 0, minutesEarlyLeave: 0 },
  //       { code: '1002', attendanceDate: new Date(2020, 11, 26),
  //         shift: 'ลากิจ', notice: 1, compensation: 1, minutesLate: 10, minutesEarlyLeave: 5 },
  //     ])
  //
  //     const attendanceMonth = toMomentObject(new Date(Date.UTC(2020, 8, 1)))
  //     const values = await getMonthlyEmployeesAttendances(attendanceMonth)
  //
  //     expect(values.map(employee => [
  //       employee.code,
  //       employee.notice,
  //       employee.vacation,
  //       employee.sickLeave,
  //       employee.casualLeave,
  //       employee.compensation,
  //       employee.minutesLate,
  //       employee.minutesEarlyLeave,
  //       employee.noShow
  //     ])).toEqual([
  //       ['1001', 2, 1, 2, 1, 2, 5, 7, 1],
  //       ['1002', 0, 0, 2, 1, 2, 0, 0, 0],
  //       ['1003', 0, 0, 0, 0, 0, 0, 0, 0],
  //     ])
  //   })
  //
  // })


})
