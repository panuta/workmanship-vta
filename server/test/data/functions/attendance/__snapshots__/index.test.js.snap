// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee Attendance functions getEmployeesAttendancesUntilAttendanceMonth function should return from all employees 1`] = `
Array [
  Object {
    "attendanceDate": "2019-12-26",
    "code": "1001",
    "compensation": 0,
    "id": 2,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-09-25",
    "code": "1001",
    "compensation": 0,
    "id": 3,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2019-12-26",
    "code": "1002",
    "compensation": 0,
    "id": 8,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-09-25",
    "code": "1002",
    "compensation": 0,
    "id": 9,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2019-12-26",
    "code": "1004",
    "compensation": 0,
    "id": 15,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
]
`;

exports[`Employee Attendance functions getEmployeesAttendancesUntilAttendanceMonth function should return from selected employees 1`] = `
Array [
  Object {
    "attendanceDate": "2019-12-26",
    "code": "1001",
    "compensation": 0,
    "id": 18,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
  Object {
    "attendanceDate": "2020-09-25",
    "code": "1001",
    "compensation": 0,
    "id": 19,
    "minutesEarlyLeave": 0,
    "minutesLate": 0,
    "notice": 0,
    "overtime": 0,
    "shift": "A",
  },
]
`;
